### PR TITLE
ref: Replace failure with std::error::Error

### PR DESCRIPTION
**Pull request by jan-auer**
Ref https://github.com/getsentry/symbolic/issues/259

### CODE REVIEW

**Review by flub — APPROVED** [review]

**Comment by jan-auer** [comment]
This is essentially ready for review, thanks a bunch to @Swatinem for wrapping this up. 

I'll also do another pass through all changes, so please don't merge just yet.

File: symbolic-cabi/src/core.rs [hunk 1]
```diff
@@ -435,9 +438,11 @@ pub unsafe extern "C" fn symbolic_err_get_last_message() -> SymbolicStr {
     use std::fmt::Write;
     LAST_ERROR.with(|e| {
         if let Some(ref err) = *e.borrow() {
-            let mut msg = err.to_string();
-            for cause in err.iter_causes() {
+            let mut err = err.as_ref();
+            let mut msg = format!("Error: {}", err);
```
Comment (jan-auer) [inline symbolic-cabi/src/core.rs:442]: Let's just use the plain `.to_string()` here.

File: symbolic-cabi/src/core.rs [hunk 2]
```diff
@@ -449,21 +454,7 @@ pub unsafe extern "C" fn symbolic_err_get_last_message() -> SymbolicStr {
...
-                write!(&mut out, "stacktrace: {}", backtrace).ok();
-                SymbolicStr::from_string(out)
-            } else {
-                Default::default()
-            }
-        } else {
-            Default::default()
-        }
-    })
+    LAST_ERROR.with(|_| Default::default())
```
Comment (jan-auer) [inline symbolic-cabi/src/core.rs:457]:
```suggestion
    SymbolicStr::default()
```

File: symbolic-debuginfo/src/elf.rs [hunk 3]
```diff
@@ -4,10 +4,11 @@ use std::borrow::Cow;
 use std::fmt;
 use std::io::Cursor;
 
-use failure::Fail;
 use flate2::{Decompress, FlushDecompress};
 use goblin::elf::compression_header::{CompressionHeader, ELFCOMPRESS_ZLIB};
-use goblin::{container::Ctx, elf, error::Error as GoblinError, strtab};
+pub use goblin::error::Error as GoblinError;
```
Comment (jan-auer) [inline symbolic-debuginfo/src/elf.rs:9]: _nit_: Lets move this under the imports as this is now a re-export. Either inherit docs or add a doc string.

File: symbolic-debuginfo/src/object.rs [hunk 4]
```diff
@@ -60,40 +60,39 @@ macro_rules! map_result {
 
 /// An error when dealing with any kind of [`Object`](enum.Object.html).
 #[non_exhaustive]
-#[allow(clippy::large_enum_variant)]
-#[derive(Debug, Fail)]
+#[derive(Debug, Error)]
```
Comment (jan-auer) [inline symbolic-debuginfo/src/object.rs:63]: On most errors we started deriving clone, except on this one. I'm considering to remove all `Clone` implementations, but let's check if we need them in Symbolicator.

File: symbolic-debuginfo/src/pdb.rs [hunk 5]
```diff
@@ -1066,7 +1046,7 @@ impl<'s> Unit<'s> {
                     let parent_offset = proc_offsets
                         .last()
                         .map(|&(_, offset)| offset)
-                        .ok_or_else(|| PdbError::from(PdbErrorKind::UnexpectedInline))?;
+                        .ok_or_else(|| PdbError::UnexpectedInline)?;
```
Comment (jan-auer) [inline symbolic-debuginfo/src/pdb.rs:1049]:
```suggestion
                        .ok_or(PdbError::UnexpectedInline)?;
```

File: symbolic-debuginfo/src/sourcebundle.rs [hunk 6]
```diff
@@ -858,7 +865,7 @@ where
     pub fn write_object<O>(self, object: &O, object_name: &str) -> Result<bool, SourceBundleError>
     where
         O: ObjectLike,
-        O::Error: Fail,
+        O::Error: Into<ObjectError>,
```
Comment (jan-auer) [inline symbolic-debuginfo/src/sourcebundle.rs:868]: I'm realizing that this is fairly restrictive now, essentially making it impossible that anyone outside of symbolic implements `ObjectLike` and then uses it with this class. I think we'll have to go back to `Box<dyn>` for this, or we simply swallow the cause.

Comment (Swatinem) [reply symbolic-debuginfo/src/sourcebundle.rs:868]: note to self: The `O::Error` constraint should be `dyn Error + Send + Sync + 'static`, and in the places where we save it its boxed.

File: symbolic-debuginfo/src/object.rs (hunk 4 above)
Comment (Swatinem) [reply symbolic-debuginfo/src/object.rs:63]: note to self: remove `Clone` on all the other custom error types…

**Review by jan-auer — COMMENTED** [review]
Looks good, let's ship this.

**Review by Swatinem — APPROVED** [review]

**Pull request merged**